### PR TITLE
Fix quantized play from the pre-roll in beatmap tracks 

### PR DESCRIPTION
**Pull request by daschuer**
This fixes https://bugs.launchpad.net/mixxx/+bug/1920084

It also fixes a bug that the user offset during sync was applied to the wrong deck.
And that quantize play syncs sometimes to the last beat rather than the next.

Tests are in place.

### CODE REVIEW

**Comment by Holzhaus** [comment]
Thanks. I've not tested this yet, but my impression was that root issue is that there ain't any beats before the track start. We should interpolate them from the beat distance between beat 1 and 2. This would also fix the issue that you can beatjump backwards from beat 1 into the Preroll.

**Comment by daschuer** [comment]
Exactly this is done here.

The beatloop issue remains. This need to be fixed in another place.

**Comment by uklotzde** [comment]
main or 2.3?

**Comment by daschuer** [comment]
Ups ... 2.3

**Comment by Holzhaus** [comment]
> Exactly this is done here.

Doesn't work for me:

Peek 2021-03-24 17-19
Left Deck: BeatMap (Preroll has no beat markers and does not support beatjumps)
Right Deck: BeatGrid (Preroll has beat markers and supports beatjumps)

I we don't have to fix it here, I'm fine merging this and then getting #2961 in a mergeable state for 2.4 ASAP.

**Comment by daschuer** [comment]
Ah, yes. I think this should be fixed by actually marking the beats in the preroll. 

This here is a solution if you have regions with no beats. I just use the next possible beat in the future for matching them.

File: src/engine/controls/bpmcontrol.cpp [hunk 1]
```diff
@@ -954,6 +969,9 @@ double BpmControl::getBeatMatchPosition(
             // loops are catching
         }
     }
+
+    qDebug() << "dNewPlaypos" << dNewPlaypos << dThisPosition;
```
Comment (uklotzde) [inline src/engine/controls/bpmcontrol.cpp:973]: Do we really need this debug output?

File: src/engine/controls/bpmcontrol.cpp [hunk 2]
```diff
@@ -880,29 +896,28 @@ double BpmControl::getBeatMatchPosition(
         return dThisPosition;
     }
 
-    double dThisSampleRate = m_pBeats->getSampleRate();
-    double dThisRateRatio = m_pRateRatio->get();
+    qDebug() << "other" << dOtherBeatFraction << dOtherPosition;
```
Comment (uklotzde) [inline src/engine/controls/bpmcontrol.cpp:899]: Please remove all those debug logs. I suppose they cause a lot of log spam?

File: src/test/enginesynctest.cpp [hunk 3]
```diff
@@ -2257,7 +2259,7 @@ TEST_F(EngineSyncTest, SeekStayInPhase) {
     ProcessBuffer();
 
     // We expect to be two buffers ahead in a beat near 0.2
-    EXPECT_DOUBLE_EQ(0.050309901738473183,
+    EXPECT_DOUBLE_EQ(0.050309901738473162,
```
Comment (uklotzde) [inline src/test/enginesynctest.cpp:2262]: The comment says "near" but we are using an equality comparison with hardcoded expected values?

File: src/engine/controls/bpmcontrol.cpp [hunk 4]
```diff
@@ -880,29 +896,28 @@ double BpmControl::getBeatMatchPosition(
...
-
-    if (dOtherBeatFraction < 1.0 / 8) {
-        // the user has probably pressed play too late, sync the previous beat
-        dOtherBeatFraction += 1.0;
+    // We can either match the past beat with dOtherBeatFraction 1.0
+    // or the next beat with dOtherBeatFraction 0.0
+    // We prefer the next because this is what will be played,
+    // unless we are close to the previous.
+    // This happens if the user presses play too late.
+    if (dOtherBeatFraction > 7 / 8.0) {
```
Comment (uklotzde) [inline src/engine/controls/bpmcontrol.cpp:906]: Define a constexpr for 7/8? Or maybe a constexpr for 1/8 is more suitable and then use `1 - kConstantValue` here?

File: src/test/enginesynctest.cpp (hunk 3 above)
Comment (daschuer) [reply src/test/enginesynctest.cpp:2262]: Yes, this value is experimental determined. The exact recalculation is missing, but that is out of scope of this PR.

**Comment by daschuer** [comment]
Done.

**Comment by Holzhaus** [comment]
> This here is a solution if you have regions with no beats. I just use the next possible beat in the future for matching them.

I don't understand what this entails. Can you make a before/after GIF?

**Comment by Holzhaus** [comment]
I think this doesn't work correctly (or I don't understand what this PR is supposed to fix). Here' s a video of the issue: https://www.youtube.com/watch?v=MKlzFQ_CpKU

**Review by uklotzde — APPROVED** [review]
@ywwg I am not familiar with the sync code.

No more concerns about the code in general.

**Comment by Holzhaus** [comment]
> I think this doesn't work correctly (or I don't understand what this PR is supposed to fix). Here' s a video of the issue: https://www.youtube.com/watch?v=MKlzFQ_CpKU

@daschuer IIUC this PR fixes a different bug that let's the track jump to the first beat? If so, LGTM.

**Comment by daschuer** [comment]
Correct

**Pull request merged**